### DISCUSSION TITLE
Better Rails routes integration

### DISCUSSION THREAD

**Discussion by renchap** [janko/rodauth-rails#37] [Ideas]
I am approaching completion of my migration from Devise to Rodauth and the fact that Rodauth routes are not proper Rails routes is a bit annoying:
- In tests (and non-controller places) you dont have access to the Rodauth route methods
- No `_url` helpers
- URL helpers do not accept parameters (if you want to add a `GET` parameter to an URL)
- They are not displayed in `rails routes`

As a side node, on the other way around Rails route helpers are not accessible from your Rodauth App, so you can not do things like this out of the box:
```
login_redirect { activity_path }
```

I think it should be possible to have a method to add these routes to `routes.rb`, reading them from Rodauth's routes and instanciating them as standard rails routes so the helpers are here. They can be routed to the `RodauthController` and not implemended here as the middleware will intercept them.

What do you think?

**Comment by janko** [comment]
Yeah, integrating with Rails routes is something I've been considering, but didn't yet find a robust way to handle that. I would definitely like to explore this more deeply, and would appreciate any assistance.

> * In tests (and non-controller places) you dont have access to the Rodauth route methods
> * No _url helpers
> * URL helpers do not accept parameters (if you want to add a GET parameter to an URL)

Note that you can use `Rodauth::Rails.rodauth` to create a Rodauth instance anywhere, including tests, which has access to the routes:

```rb
rodauth = Rodauth::Rails.rodauth
rodauth.login_path # *_path methods
rodauth.logout_url # *_url_methods
rodauth.create_account_path(type: "admin") # GET query parameters
```

> * They are not displayed in rails routes

I wish Rails had some API for registering custom routes. I was looking into somehow extending the `rails routes` task, but it proved to be too complicated. For now you have the `rodauth:routes` Rake task, but I agree it's not the same.

> As a side node, on the other way around Rails route helpers are not accessible from your Rodauth App

You can currently do access them via `Rails.application.routes.url_helpers`:

```rb
login_redirect { Rails.application.routes.url_helpers.activity_path }
```

I thought about `include Rails.application.routes.url_helpers` into the Rodauth class, but I worry about method name clashes.

> I think it should be possible to have a method to add these routes to routes.rb, reading them from Rodauth's routes and instanciating them as standard rails routes so the helpers are here. 

This would also probably automatically display them in `rails routes`. One of the challenges I remember was naming the Rails routes, because you can have `rodauth.login_path`, or `rodauth(:admin).login_path` if you're using multiple Rodauth configurations. Maybe something like `login_path(:admin)`.

> They can be routed to the RodauthController and not implemented here as the middleware will intercept them.

Note that it's not possible that RodauthController handles those routes, because the endpoint functionality is implemented on the Roda routing level.

**Reply by renchap** [reply]
> Note that you can use `Rodauth::Rails.rodauth` to create a Rodauth instance anywhere, including tests, which has access to the routes:
> 
> ```ruby
> rodauth = Rodauth::Rails.rodauth
> rodauth.login_path # *_path methods
> rodauth.logout_url # *_url_methods
> rodauth.create_account_path(type: "admin") # GET query parameters
> ```

Good point, this would be better than my current helpers with static routes. 

> > As a side node, on the other way around Rails route helpers are not accessible from your Rodauth App
> 
> You can currently do access them via `Rails.application.routes.url_helpers`:
> 
> ```ruby
> login_redirect { Rails.application.routes.url_helpers.activity_path }
> ```
> 
> I thought about `include Rails.application.routes.url_helpers` into the Rodauth class, but I worry about method name clashes.

Yes, especially if Rodauth routes are somehow integrated into Rails routes. But maybe a helper that can be used as `rails_route.activity_path`?

> This would also probably automatically display them in `rails routes`. One of the challenges I remember was naming the Rails routes, because you can have `rodauth.login_path`, or `rodauth(:admin).login_path` if you're using multiple Rodauth configurations. Maybe something like `login_path(:admin)`.

I would use a prefix to indicate that those routes are handled by rodauth. Maybe something like:

```
rodauth.login_path => rodauth_login_path
rodauth(:admin).login_path => rodauth_admin_login_path
```

Maybe even make the `rodauth_` prefix configurable, so people can use `user_` for their default config, and `user_admin_` for their admin config if they prefer?

> > They can be routed to the RodauthController and not implemented here as the middleware will intercept them.
> 
> Note that it's not possible that RodauthController handles those routes, because the endpoint functionality is implemented on the Roda routing level.

Yes I know, but iirc your Rails route needs to be routed somewhere (`Controller#Action` column in `rails route`), so you could route them to `RodauthController` and have a comment in this controller saying that the routes are intercepted by Roda and needs to be defined in the Rodauth app instead.

**Reply by janko** [reply]
> But maybe a helper that can be used as rails_route.activity_path?

Just added the `#rails_routes` method to master – https://github.com/janko/rodauth-rails/commit/c00e43f9818b89108c0795420b331b786362dc8d.

> Maybe even make the rodauth_ prefix configurable, so people can use user_ for their default config, and user_admin_ for their admin config if they prefer?

Yeah, it would definitely be nice to allow the user to choose the prefixes. I now remembered that one of the challenges with defining Rails routes was handling reloading in development; if the developer enables a new feature, we should ensure the new route is dynamically defined (and routes for removed features undefined).

> iirc your Rails route needs to be routed somewhere (Controller#Action column in rails route), so you could route them to RodauthController and have a comment in this controller saying that the routes are intercepted by Roda and needs to be defined in the Rodauth app instead.

I don't think it's possible to only visually display Rodauth routes being routed to the `RodauthController` in the `rails routes` output, without this actually being true. From my previous research, the API behind `rails routes` isn't extendable like that.

**Reply by renchap** [reply]
> > But maybe a helper that can be used as rails_route.activity_path?
> 
> Just added the `#rails_routes` method to master – [c00e43f](https://github.com/janko/rodauth-rails/commit/c00e43f9818b89108c0795420b331b786362dc8d).

👍 

> 
> > Maybe even make the rodauth_ prefix configurable, so people can use user_ for their default config, and user_admin_ for their admin config if they prefer?
> 
> Yeah, it would definitely be nice to allow the user to choose the prefixes. I now remembered that one of the challenges with defining Rails routes was handling reloading in development; if the developer enables a new feature, we should ensure the new route is dynamically defined (and routes for removed features undefined).

Yes reloading is definitely an issue. I never looked at how it works with rails and if there is a way to force reloading the `routes.rb` when the rodauth app (or one of its dependencies) change.

> I don't think it's possible to only visually display Rodauth routes being routed to the `RodauthController` in the `rails routes` output, without this actually being true. From my previous research, the API behind `rails routes` isn't extendable like that.

As Roda routes will be processed before Rails ones, you can tell to Rails to route it but in practice it will be intercepted before by Roda and Rails routing wont matter, but at least the route will be displayed with a coherent controller in `rails route`.

For example, if I do:

```ruby
# routes.rb
get "/auth/login", to: "rodauth#login"

# rodauth_controller.rb
class RodauthController < ApplicationController
  # This controller is a placeholder, Rodauth routes are handled by Roda and wont trigger actions here

  def login
     logger.info "Login!" 
  end
end
```

Then a request to `/auth/login` wont be hitting my `RodauthController#login` method and will correctly be processed by Rodauth, but `rails route` shows it:

```
$ rails routes | grep auth  
auth_login GET     /auth/login(.:format)                 rodauth#login
```

**Reply by janko** [reply]
> As Roda routes will be processed before Rails ones, you can tell to Rails to route it but in practice it will be intercepted before by Roda and Rails routing wont matter, but at least the route will be displayed with a coherent controller in rails route.

You're right, that could work then. It might be a bit misleading, given that `RodauthController` won't actually handle these requests, but it's probably still better than not displaying any routes in `rails routes`.

Thank you for giving me food for thought, I'll explore this in more detail.

**Reply by janko** [reply]
Just tested the [`#direct`](https://api.rubyonrails.org/classes/ActionDispatch/Routing/Mapper/CustomUrls.html#method-i-direct) directive, and the following declaration correctly generates `#login_path` and `#login_url` helpers for relative and absolute paths respectively, and GET query parameters work as well:

```rb
  direct :login do |options = {}|
    rodauth = Rodauth::Rails.rodauth
    rodauth.scope.env.merge!(request.env)
    rodauth.login_url(options)
  end
```

However, as expected, they don't appear in `rails routes`. I still feel reluctant to add dummy route declaration that point to Rodauth actions that don't exist, I will give that some more thought. It might be worth the tradeoff, though.

**Reply by renchap** [reply]
This looks cleaner that what I am doing, I will try to switch to it.

Currently this is what I deployed, working fine but I should probably fetch the routes list from rodauth rather than listing them manually

```ruby
module RodauthUrlHelpers
  %i[login logout reset_password reset_password_request change_password create_account verify_account
     change_login verify_login_change].each do |method|
    %i[path url].each do |suffix|
      define_method "user_#{method}_#{suffix}" do |*args, **options|
        I18n.with_locale(options[:locale]) do
          Rodauth::Rails.rodauth.send("#{method}_#{suffix}", *args, **options.except(:locale))
        end
      end
    end
  end
end

```

**Comment by janko** [comment]
Two and a half years later, I finally realized the importance of having native URL helpers for Rodauth routes you were advocating for. In my case, I was porting rubygems.org authentication to Rodauth just as an experiment, and the locale switcher that used `url_for(locale: locale)` didn't work on Rodauth pages. I could manually implement the logic for Rodauth pages, but I didn't want to think, I just wanted it to work 😃 

I implemented this in https://github.com/janko/rodauth-rails/pull/245. If you have some time, would you mind giving it a try? It should just work™️ with locale prefixes as well:

```rb
Rails.application.routes.draw do
  scope "(:locale)" do
    rodauth
  end
end
```
```rb
create_account_path(locale: "hr") #=> "/hr/create-account"
```
